designs/0005-static-matrices.md

```diff
@@ -0,0 +1,95 @@
+- Start Date: April 30th, 2020
+- RFC PR: [Example Branch](https://github.com/stan-dev/math/compare/feature/var-template)
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes a `static_matrix` and `static_vector` type in the Stan language which on the backend will allow for significant performance opportunities. In Stan Math this will be represented by `var<Eigen::Matrix<double, 1, 1>>` or `var<matrix_cl<double>>`. The implementation will be fully backwards compatible with current Stan code. With optimizations from the compiler the conditions for a `matrix` to be a `static_matrix` can be detected and applied to Stan programs automatically for users.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, a differentiable matrix in Stan is represented as an Eigen matrix holding a pointer to an underlying array of autodiff objects, each of those holding a pointer to the underlying autodiff object implementation. These `N*M` autodiff elements are expensive but necessary so the elements of a matrix can be assigned to without copying the entire matrix. However, in instances where the matrix is treated as a whole object such that underlying subsets of the matrix are never assigned to, we can store one autodiff object representing the entire matrix. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+At the language level, `static_matrix` can replace a `matrix` type if subsets of the matrix are never assigned to. That makes code such as the below illegal.
+
+```stan
+static_matrix[N, M] A = // Fill in A...
+A[10, 10] = 5;
+```
+
+Besides that they can be thought of as standard matrix.
+
+At the Stan Math level a `static_matrix` is a `var_type<Eigen::Matrix<double, -1, -1>>` with an underlying pointer to a `vari_type<Eigen::Matrix<double, -1, -1>>`*. This means that accessors for the value and adjoint of `var_type` objects can access contiguous chunks of each part of the `var_type`. Any function that accept a `var_type<T>` will support static matrices.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature will be added in the following stages.
+
+1. Replace `var` and `vari` with templated types. See the [example branch](https://github.com/stan-dev/math/compare/feature/var-template) here to see how this will be done. `var` and `vari` have been changed to `var_type<T>` and `vari_type<T>` with aliases for the current methods as `using var = var_type<double>` and `using vari = vari_type<double>`. These aliases allow for full backwards compatibility with existing upstream dependencies.
+
+2. Make the stack allocator conform to accept `vari_type<T>` objects.
+
+3. Additional PRs for the current arithmetic operators for `var` to accept `vari_type<T>` object types.
+
+4. Sort out and add any additional specialized methods for static matrix types such as `cholesky_decompose`.
```

Comment:
What does this mean?  Is every matrix function going to have to be rewritten or will there be a general, perhaps slow, way to get this off the ground quickly?

Reply:
> What does this mean? 

Any thing in (4) is an optional method we would want to use with a sparse matrix. Like for Cholesky we would want

```cpp
var_type<Eigen<double, -1, -1>> cholesky_decompose(var_type<Eigen<double, -1, -1>>)
```

and a reverse mode specialization that takes advantage of the structure

> Is every matrix function going to have to be rewritten or will there be a general, perhaps slow, way to get this off the ground quickly?

Let me think about this, I need to reread the issue as @t4c1 talked about a specialization of `Eigen::Matrix<var>` there that might save us from rewriting a lot of code

Reply:
@SteveBronder Can you maybe determine in advance which functions need to be duplicated for static matrix and which can remain same?

Reply:
> @SteveBronder Can you maybe determine in advance which functions need to be duplicated for static matrix and which can remain same?

I think the answer is anything that takes in `Eigen::Matrix<var>` right now in the constructor for the vari class that holds the `chain()` method we do something like making a `vari**` member that then adds N `vari`s to the stack. Unless you know a simpler scheme we'd need to run through all those and make ones that construct two `T**` members (with T being the eigen scalar type). But with the way this is setup we can slowly do those. We can also delay releasing this as a language feature until we have everything supported that `matrix` has while still telling the compiler what signatures support sparse matrices so it can run through the program and exchange `matrix` for `static_matrix` where appropriate.